ValveResourceFormat/KeyValues/KVObject.cs

```diff
@@ -0,0 +1,193 @@
+
+                stringBuilder.Append("\n");
+            }
+
+            for (var i = 0; i < indent - 1; i++)
+            {
+                stringBuilder.Append("\t");
+            }
+            stringBuilder.Append("}");
+        }
+
+        private void SerializeArray(StringBuilder stringBuilder, int indent)
+        {
+            stringBuilder.Append("\n");
+            for (var i = 0; i < indent - 1; i++)
+            {
+                stringBuilder.Append("\t");
+            }
+            stringBuilder.Append("[\n");
+
+            //Need to preserve the order
+            for (var i = 0; i < Count; i++)
+            {
+                for (var j = 0; j < indent; j++)
+                {
+                    stringBuilder.Append("\t");
+                }
+
+                PrintValue(stringBuilder, Properties[i.ToString()], indent);
+
+                stringBuilder.Append(",\n");
+            }
+
+            for (var i = 0; i < indent - 1; i++)
+            {
+                stringBuilder.Append("\t");
+            }
+            stringBuilder.AppendLine("]");
+        }
+
+        //Print a value in the correct representation
+        private void PrintValue(StringBuilder stringBuilder, KVValue kvValue, int indent)
+        {
+            KVType type = kvValue.Type;
+            object value = kvValue.Value;
+
+            switch (type)
+            {
+                case KVType.OBJECT:
+                    ((KVObject)value).Serialize(stringBuilder, indent + 1);
+                    break;
+                case KVType.ARRAY:
+                    ((KVObject)value).Serialize(stringBuilder, indent + 1);
+                    break;
+                case KVType.FLAGGED_STRING:
+                    stringBuilder.Append((string)value);
+                    break;
+                case KVType.STRING:
+                    stringBuilder.Append("\"");
+                    stringBuilder.Append((string)value);
+                    stringBuilder.Append("\"");
+                    break;
+                case KVType.STRING_MULTI:
+                    stringBuilder.Append("\"\"\"\n");
+                    stringBuilder.Append((string)value);
+                    stringBuilder.Append("\n\"\"\"");
+                    break;
+                case KVType.BOOLEAN:
+                    stringBuilder.Append((bool)value ? "true" : "false");
+                    break;
+                case KVType.DOUBLE:
+                    stringBuilder.Append(((double)value).ToString("#0.000000", CultureInfo.InvariantCulture));
+                    break;
+                case KVType.INTEGER:
+                    stringBuilder.Append((int)value);
+                    break;
+                default:
+                    //Unknown type encountered
+                    throw new Exception("Unknown type encountered.");
+            }
+        }
+    }
+
+    //Class to hold type + value
+    public class KVValue
```

Comment:
For consideration: Should `KVValue` be mutable, or immutable? If it should be immutable, make `Type` and `Value` read-only properties.

Also for consideration: As such a simple type, can `KVValue` be a struct, or must it be an object?